lib/line-numbers.js

```diff
@@ -0,0 +1,148 @@
+	})
+)));
+
+function splitPatternAndLineNumbers(pattern) {
+	const parts = pattern.split(DELIMITER);
+	if (parts.length === 1) {
+		return {pattern, lineNumbers: null};
+	}
+
+	const suffix = removeAllWhitespace(parts.pop());
+	if (!LINE_NUMBERS_REGEX.test(suffix)) {
+		return {pattern, lineNumbers: null};
+	}
+
+	return {pattern: parts.join(DELIMITER), lineNumbers: parseLineNumbers(suffix)};
+}
+
+exports.splitPatternAndLineNumbers = splitPatternAndLineNumbers;
+
+function getApplicableLineNumbers(normalizedFilePath, filter) {
+	return sortNumbersAscending(distinctArray(flatten(
+		filter
+			.filter(({pattern, lineNumbers}) => lineNumbers && micromatch.isMatch(normalizedFilePath, pattern))
+			.map(({lineNumbers}) => lineNumbers)
+	)));
+}
+
+exports.getApplicableLineNumbers = getApplicableLineNumbers;
+
+function initializeLineNumberSelector(file, lineNumbers = []) {
+	if (lineNumbers.length === 0) {
+		return undefined;
+	}
+
+	// Load lazily, just in the worker processes.
+	const fs = require('fs');
+	const recast = require('recast');
+
+	// Fail silently if the file is missing (which would be really surprising)
+	// or can't be parsed, perhaps because it's a TypeScript file.
+	let ast;
+	try {
+		ast = recast.parse(fs.readFileSync(file, 'utf8'));
+	} catch {
+		return undefined;
+	}
+
+	// Record the start and end positions of all call expressions.
+	const allCallLocations = [];
+	recast.visit(ast, {
+		visitCallExpression(path) {
+			const {start, end} = path.node.loc;
+			allCallLocations.push({start, end});
+			this.traverse(path);
+		}
+	});
+
+	const selectedLines = new Set(lineNumbers);
+
+	return {
+		isSelected() {
+			// Assume this is called from a test declaration, which is located in the file.
+			// If not… don't select the test!
+			const callSite = callsites().find(callSite => callSite.getFileName() === file);
+			if (!callSite) {
+				return false;
+			}
+
+			const sourceCallSite = sourceMapSupport.wrapCallSite(callSite);
+			const start = {
+				line: sourceCallSite.getLineNumber(),
+				column: sourceCallSite.getColumnNumber() - 1 // Use 0-indexed columns.
+			};
+
+			// Find all calls that close over the test declaration.
+			const enclosingCalls = allCallLocations.filter(loc => {
+				if (loc.start.line > start.line || loc.end.line < start.line) {
+					return false;
+				}
+
+				if (loc.start.line === start.line && loc.start.column > start.column) {
+					return false;
+				}
+
+				if (loc.end.line === start.line && loc.end.column < start.column) {
+					return false;
+				}
+
+				return true;
+			});
+
+			// Call locations should be sorted by source order, so the last enclosing
+			// call must be the test declaration.
+			if (enclosingCalls.length === 0) {
+				return false;
+			}
+
+			const testDeclaration = enclosingCalls.pop();
+			for (let {line} = testDeclaration.start; line <= testDeclaration.end.line; line++) {
+				if (selectedLines.has(line)) {
+					return true;
+				}
+			}
+
+			return false;
```

Comment:
I would still prefer:

```suggestion
			return range(testDeclaration.start.line, testDeclaration.end.line).some(line => selectedLines.has(line));
```

Reply:
Went ahead and did this. More dense and easier to read IMO.